profile/filter_test.go

```diff
@@ -0,0 +1,252 @@
+		{ID: 6, Mapping: mappings[0], Address: 0x6000, Line: []Line{{Function: functions[5], Line: 1}}},
+		{ID: 7, Mapping: mappings[0], Address: 0x7000, Line: []Line{{Function: functions[6], Line: 1}}},
+		{ID: 8, Mapping: mappings[0], Address: 0x8000, Line: []Line{{Function: functions[7], Line: 1}}},
+		{ID: 9, Mapping: mappings[0], Address: 0x9000, Line: []Line{{Function: functions[8], Line: 1}}},
+		{ID: 10, Mapping: mappings[0], Address: 0x10000, Line: []Line{{Function: functions[9], Line: 1}}},
+		{ID: 11, Mapping: mappings[1], Address: 0x11000, Line: []Line{{Function: functions[10], Line: 1}}},
+	}
+	profile := &Profile{
+		TimeNanos:     10000,
+		PeriodType:    &ValueType{Type: "cpu", Unit: "milliseconds"},
+		Period:        1,
+		DurationNanos: 10e9,
+		SampleType: []*ValueType{{Type: "samples", Unit: "count"}},
+		Mapping:  mappings,
+		Function: functions,
+		Location: locations,
+		Sample: []*Sample{
+			{Value: []int64{1}, Location: []*Location{locations[0], locations[1], locations[2], locations[3]}},
+			{Value: []int64{2}, Location: []*Location{locations[4], locations[5], locations[1], locations[6]}},
+			{Value: []int64{3}, Location: []*Location{locations[7], locations[8]}},
+			{Value: []int64{4}, Location: []*Location{locations[9], locations[4], locations[10], locations[7]}},
+		},
+	}
+
+	for _, testCase := range []struct{
+		name string
+		focus, ignore, hide, show *regexp.Regexp
+		// todo: fm, im, hm, hnm           bool
+		wantPaths [][]int // expected paths. wantPaths[sampleIndex][frameIndex] = locationIndex
+		wantValues []int64 // expected values. wantValues[sampleIndex] = sampleValue
+	}{
+		{
+			name: "empty filters keep all frames",
+			// todo: fm: true, // nil focus matches every sample
+			wantPaths: [][]int{
+				{0, 1, 2, 3},
+				{4, 5, 1, 6},
+				{7, 8},
+				{9, 4, 10, 7},
+			},
+			wantValues: []int64{1, 2, 3, 4},
+		},
+		{
+			name: "hides a function if both show and hide match it",
+			show: regexp.MustCompile("fun1"),
+			hide: regexp.MustCompile("fun10"),
+			wantPaths: [][]int{
+				{1},
+				{1},
+			},
+			wantValues: []int64{1, 2},
+		},
+	} {
+		t.Run(testCase.name, func(t *testing.T) {
+			gotProfile := profile.Copy()
+			gotProfile.FilterSamplesByName(testCase.focus, testCase.ignore, testCase.hide, testCase.show)
+			gotStr := gotProfile.String()
+
+			wantSamples, wantLocations := makeFilteredSamples(testCase.wantPaths, testCase.wantValues, locations)
+			wantProfile := profile.Copy()
+			wantProfile.Sample = wantSamples
+			wantProfile.Location = wantLocations
+			wantStr := wantProfile.String()
+
+			if wantStr != gotStr {
+				diff, err := proftest.Diff([]byte(wantStr), []byte(gotStr))
+				if err != nil {
+					t.Fatalf("failed to get diff: %v", err)
+				}
+				t.Errorf("filtered profile got diff(want->got): \n%s", diff)
+			}
+		})
+	}
+}
+
+// makeSamples constructs profile samples from basic inputs.
+// paths: location indices, indexed by sample and frame.
```

Comment:
Nit: I don't remember seeing such per-argument format of documentation in Go code or pprof, could you look at existing documentation for functions in pprof and document similarly?

Reply:
i looked and couldnt find anything, but i think this is right thing to do here.

Reply:
Try harder :)

Reply:
Go doc style is that the text should be readable English, not some kind of machine-friendly special formatting. One example is https://github.com/golang/go/blob/e13a213c7f9d78cfb79f91bcf9b75e3d22362a92/src/runtime/map.go#L691, there are more around.